doc/man/doveadm-kick.1.in
.\" Copyright (c) 2010-2018 Dovecot authors, see the included COPYING file
.TH DOVEADM\-KICK 1 "2010-06-12" "Dovecot v2.3" "Dovecot"
.SH NAME
doveadm\-kick \- Disconnect users by user name and/or IP address
.\"------------------------------------------------------------------------
.SH SYNOPSIS
.BR doveadm " [" \-Dv "] " kick "
.RB [ -a
.IR anvil_socket_path ]
.RB [ \-f
.IR passdb_field ]
.RB [ \-h
.IR dest_host ]
.I user_mask
.\"-------------------------------------
.br
.BR doveadm " [" \-Dv "] " kick "
.RB [ -a
.IR anvil_socket_path ]
.RB [ \-f
.IR passdb_field ]
.RB [ \-h
.IR dest_host ]
\fIip\fP[\fB/\fP\fIbits\fP]
.\"-------------------------------------
.br
.BR doveadm " [" \-Dv "] " kick "
.RB [ -a
.IR anvil_socket_path ]
.RB [ \-f
.IR passdb_field ]
.RB [ \-h
.IR dest_host ]
.I user_mask
\fIip\fP[\fB/\fP\fIbits\fP]
.\"-------------------------------------
.br
.BR doveadm " [" \-Dv "] " kick "
.RB [ -a
.IR anvil_socket_path ]
.RB [ \-f
.IR passdb_field ]
.B \-h
.I dest_host
.\"------------------------------------------------------------------------
.SH DESCRIPTION
.BR doveadm \(aqs\  kick
command is used to disconnect users by
.I user_mask
and/or the
.I ip
address, from which they are connected.
.PP
In the first form, all users, whose login name matches the
.I user_mask
argument, will be disconnected.
.PP
In the second form, all users, connected from the given IP address or
network range, will be disconnected.
.PP
In the third form, only users connected from the given IP address or
networks range and a matching login name will be disconnected.
.PP
In the last form, all proxy connections to the given destination host are
disconnected.
.\"------------------------------------------------------------------------
@INCLUDE:global-options@
.\" --- command specific options --- "/.
.PP
Command specific
.IR options :
.\"-------------------------------------
.TP
.BI \-a\  anvil_socket_path
This option is used to specify an absolute path to an alternative UNIX
domain socket.
.sp
By default
.BR doveadm (1)
will use the socket
.IR @rundir@/anvil .
The socket may be located in another directory, when the default
.I base_dir
setting was overridden in
.IR @pkgsysconfdir@/dovecot.conf .
.\"-------------------------------------
.TP
.BI \-f\  passdb_field
Alternative username field to use for kicking, as returned by passdb.
Only the passdb fields beginning with the
.I user_
prefix are tracked.
.\"-------------------------------------
.TP
.BI \-h\  dest_host
Disconnect proxy connections to the given
.IR dest_host .
.\"------------------------------------------------------------------------
.SH ARGUMENTS
.TP
.IR ip [/ bits ]
.I ip
or
.IB ip /\c
.I bits
is the host or network, from which the users are connected.
.\"-------------------------------------
.TP
.I user_mask
Is a user\(aqs login name, or the altenative username (user_* field) if the
.B -f
parameter is used.
Depending on the configuration, a login name may be for example
.BR jane " or " john@example.com .
It\(aqs also possible to use
.RB \(aq * \(aq
and
.RB \(aq ? \(aq
wildcards (e.g. \-u *@example.org).
.\"------------------------------------------------------------------------
.SH EXAMPLE
If you don\(aqt want to disconnect all users at once, you can check
who\(aqs currently logged in.
The first example demonstrates how to disconnect all users whose login name
is 3 characters long and begins with
.BR ba .
.sp
.nf
.B doveadm who \-1 ja*
username                      service pid ip
jane                          imap   8192 ::1
jano                          imap   8196 ::2
james                         imap   8203 2001:db8:543:2::1
.B doveadm kick jan?
The connections for jane and jano are kicked.
.fi
.PP
The next example shows how to kick user foo\(aqs connections from 192.0.2.*.
.sp
.nf
.B doveadm who \-1 foo
username                     service pid ip
foo                          imap   8135 fd95:4eed:38ba::25
foo                          imap   9112 192.0.2.53
foo                          imap   8216 192.0.2.111
.B doveadm kick foo 192.0.2.0/24

.B doveadm who f*
username                  # service (pids) (ips)
foo                       1 imap    (8135) (fd95:4eed:38ba::25)
.fi
.\"------------------------------------------------------------------------
@INCLUDE:reporting-bugs@
.\"------------------------------------------------------------------------
.SH SEE ALSO
.BR doveadm (1),
.BR doveadm\-who (1)

NEWS
v2.3.18 2022-02-03  Aki Tuomi <aki.tuomi@open-xchange.com>

	* Removed mail_cache_lookup_finished event. This event wasn't especially
	  useful, but it increased CPU usage significantly.
	* fts: Don't index inline base64 encoded content in FTS indexes using
	  the generic tokenizer. This reduces the FTS index sizes by removing
	  input that is very unlikely to be searched for. See
	  https://doc.dovecot.org/configuration_manual/fts/tokenization for
	  details on how base64 is detected. Only applies when using libfts.
	* lmtp: Session IDs are now preserved through proxied connections, so
	  LMTP sessions can be tracked. This slightly changes the LMTP session
	  ID format by appending ":Tn" (transaction), ":Pn" (proxy connection)
	  and ":Rn" (recipient) counters after the session ID prefix.
	+ Events now have "reason_code" field, which can provide a list of
	  reasons why the event is happening. See
	  https://doc.dovecot.org/admin_manual/event_reasons/
	+ New events are added. See https://doc.dovecot.org/admin_manual/list_of_events/
	+ fts: Added fts_header_excludes and fts_header_includes settings to
	  specify which headers to index. See
	  https://doc.dovecot.org/settings/plugin/fts-plugin#plugin-fts-setting-fts-header-excludes
	  for configuration details.
	+ fts: Initialize the textcat language detection library only once per
	  process. This can reduce CPU usage if fts_languages setting has multiple
	  languages listed and service indexer-worker { service_count } isn't 1.
	  Only applies when using libfts.
	+ lib-storage: Reduced CPU usage significantly for some operations that
	  accessed lots of emails (e.g. fetching all flags in a folder, SORT, ...)
	+ lib: DOVECOT_PREREQ() - Add micro version which enables compiling
	  external plugins against different versions of Dovecot.
	+ lmtp: Added new lmtp_verbose_replies setting that makes errors sent to
	  the LMTP client much more verbose with details about why exactly
	  backend proxy connections or commands are failing.
	+ submission: Support implicit SASL EXTERNAL with
	  submission_client_workarounds=implicit-auth-external. This allows
	  automatically logging in when SSL client certificate is present.
	- *-login: Statistics were disabled if stats process connection was lost.
	- auth: Authentication master user login fails with SCRAM-* SASL mechanisms.
	- auth: With auth_cache_verify_password_with_worker=yes, passdb extra
	  fields in the auth cache got lost.
	- doveadm: Fixed crash if zlib_save_level setting was specified,
	  but zlib_save was unset. v2.3.15 regression.
	- doveadm: Proxying can panic when flushing print output. v2.3.17
	  regression. Fixes:
	  Panic: file ioloop.c: line 865 (io_loop_destroy): assertion failed:
	  (ioloop == current_ioloop)
	- doveadm: stats add --group-by parameter didn't work.
	- fts: Using email-address fts tokenizer could result in excessive memory
	  usage with garbage email input. This could cause the indexer-worker
	  processes to fail due to reaching the VSZ memory size limit.
	  Only applies when using libfts.
	- imap: A SEARCH command timing out while fts returns indexes may timeout
	  returning "NO [SERVERBUG]", while it should return "NO [INUSE]" instead.
	- imap: LIST-EXTENDED doesn't return STATUS for all folders. Sending
	  LIST .. RETURN (SUBSCRIBED STATUS (...)) did not return STATUS for
	  folders that are not subscribed when they have a child folder that is
	  subscribed as mandated by IMAP RFCs.
	- imapc: Mailbox vsize calculation crashed with
	  Panic: file index-mailbox-size.c: line 344 (index_mailbox_vsize_hdr_add_missing):
	  assertion failed: (mails_left > 0)
	- indexer: If indexer-worker crashes, the request it was processing gets
	  stuck in the indexer process. This stops indexing for the folder until
	  indexer process is restarted. v2.3.14 regression.
	- indexer: Process was slowly leaking memory for each indexing request.
	- lib-event: Unnamed events were wrongly filtered out for event/metric
	  filters like "event=abc OR something_independent_of_event_name".
	- lib-index: 64-bit big endian CPUs handle last_used field in
	  dovecot.index.cache wrong.
	- lib-ssl-iostream: Fix buggy OpenSSL error handling without assert-crashing.
	  If there is no error available, log it as an error instead of crashing.
	  The previous fix for this in v2.3.11 was incomplete. Fixes
	  Panic: file istream-openssl.c: line 51 (i_stream_ssl_read_real):
	  assertion failed: (errno != 0)
	- lmtp: Out-of-memory issues can happen when proxying large messages to
	  LMTP backend servers that accept the message data too slow.
	- master: HAProxy header parsing has read buffer overflow if provided
	  header size is invalid. This happens only if inet_listener
	  { haproxy=yes } is configured and only if the remote IP address is in
	  haproxy_trusted_networks.
	- old_stats: Plugin kept increasing memory usage, which became
	  noticeable with long-running imap sessions.
	- stats: Dynamically adding same metric multiple times causes multiple stats.
	- submission-login: Authentication does not accept OAUTH2 token (or
	  other very long credentials) because it considers the line to be too long.
	- submission-login: Process can crash if HELO is pipelined with an
	  invalid domain.
	- submission-proxy: Don't use SASL-IR if it would make the AUTH command
	  line longer than 512 bytes.
	- submission: Service would crash if relay server authentication failed.
	- virtual: FTS search in a virtual folder could crash if there are
	  duplicate mailbox GUIDs. This mainly happened when user had both INBOX
	  and INBOX/INBOX folders and the namespace prefix was INBOX/. Fixes
	  Panic: file hash.c: line 252 (hash_table_insert_node):
	  assertion failed: (opcode == HASH_TABLE_OP_UPDATE)
	- virtual: If mailbox opening fails, the backend mailbox is leaked and
	  process crashes when client disconnects. Fixes
	  Panic: file mail-user.c: line 232 (mail_user_deinit):
	  assertion failed: ((*user)->refcount == 1)

src/doveadm/doveadm-kick.c
/* Copyright (c) 2010-2018 Dovecot authors, see the included COPYING file */

#include "lib.h"
#include "array.h"
#include "net.h"
#include "str.h"
#include "ioloop.h"
#include "strescape.h"
#include "anvil-client.h"
#include "doveadm.h"
#include "doveadm-who.h"
#include "doveadm-print.h"

#include <stdio.h>

struct kick_session {
	const char *username;
	guid_128_t conn_guid;
};

struct kick_context {
	struct who_context who;
	enum doveadm_client_type conn_type;
	ARRAY(struct kick_session) kicks;
	unsigned int kicked_count;
};

static void
kick_user_anvil_callback(const char *reply, struct kick_context *ctx)
{
	unsigned int count;

	if (reply != NULL) {
		if (str_to_uint(reply, &count) < 0)
			i_error("Unexpected reply from anvil: %s", reply);
		else
			ctx->kicked_count += count;
	}
	io_loop_stop(current_ioloop);
}

static void kick_users_get_via_who(struct kick_context *ctx)
{
	/* get a list of all user+sessions matching the filter */
	p_array_init(&ctx->kicks, ctx->who.pool, 64);
	struct doveadm_who_iter *iter =
		doveadm_who_iter_init(ctx->who.anvil_path);
	if (!doveadm_who_iter_init_filter(iter, &ctx->who.filter)) {
		doveadm_who_iter_deinit(&iter);
		return;
	}
	struct who_line who_line;
	while (doveadm_who_iter_next(iter, &who_line)) {
		if (!who_line_filter_match(&who_line, &ctx->who.filter))
			continue;
		struct kick_session *session = array_append_space(&ctx->kicks);
		session->username = p_strdup(ctx->who.pool, who_line.username);
		guid_128_copy(session->conn_guid, who_line.conn_guid);
	}
	if (doveadm_who_iter_deinit(&iter) < 0)
		doveadm_exit_code = EX_TEMPFAIL;
}

static void kick_users_via_anvil(struct kick_context *ctx)
{
	const struct kick_session *session;
	string_t *cmd = t_str_new(128);

	struct anvil_client *anvil =
		anvil_client_init(ctx->who.anvil_path, NULL, 0);
	if (anvil_client_connect(anvil, TRUE) < 0) {
		doveadm_exit_code = EX_TEMPFAIL;
		return;
	}

	array_foreach(&ctx->kicks, session) {
		str_truncate(cmd, 0);
		if (ctx->who.filter.alt_username_field == NULL) {
			str_append(cmd, "KICK-USER\t");
			str_append_tabescaped(cmd, session->username);
			if (!guid_128_is_empty(session->conn_guid)) {
				str_append_c(cmd, '\t');
				str_append_tabescaped(cmd,
					guid_128_to_string(session->conn_guid));
			}
		} else {
			str_append(cmd, "KICK-ALT-USER\t");
			str_append_tabescaped(cmd, ctx->who.filter.alt_username_field);
			str_append_c(cmd, '\t');
			str_append_tabescaped(cmd, session->username);
		}

		anvil_client_query(anvil, str_c(cmd),
				   ANVIL_DEFAULT_KICK_TIMEOUT_MSECS,
				   kick_user_anvil_callback, ctx);
		io_loop_run(current_ioloop);
	}
	anvil_client_deinit(&anvil);

	doveadm_print(dec2str(ctx->kicked_count));
}

static void cmd_kick(struct doveadm_cmd_context *cctx)
{
	const char *passdb_field, *dest_host, *const *masks = NULL;
	struct kick_context ctx;
	struct ip_addr dest_ip;

	i_zero(&ctx);
	if (!doveadm_cmd_param_str(cctx, "socket-path", &(ctx.who.anvil_path)))
		ctx.who.anvil_path = t_strconcat(doveadm_settings->base_dir, "/anvil", NULL);
	if (!doveadm_cmd_param_str(cctx, "passdb-field", &passdb_field))
		passdb_field = NULL;

	if (!doveadm_cmd_param_str(cctx, "dest-host", &dest_host))
		i_zero(&dest_ip);
	else if (net_addr2ip(dest_host, &dest_ip) < 0)
		i_fatal("dest-host isn't a valid IP address");

	if (!doveadm_cmd_param_array(cctx, "mask", &masks) &&
	    dest_ip.family == 0) {
		help_ver2(&doveadm_cmd_kick_ver2);
		return;
	}
	ctx.conn_type = cctx->conn_type;
	ctx.who.pool = pool_alloconly_create("kick pids", 10240);

	if (who_parse_args(&ctx.who, passdb_field, &dest_ip, masks) != 0) {
		pool_unref(&ctx.who.pool);
		return;
	}

	doveadm_print_init(DOVEADM_PRINT_TYPE_FORMATTED);
	doveadm_print_formatted_set_format("%{count} connections kicked\n");
	doveadm_print_header_simple("count");

	if (ctx.who.filter.net_bits == 0 &&
	    ctx.who.filter.dest_ip.family == 0 &&
	    strpbrk(ctx.who.filter.username, "*?") == NULL) {
		/* kick a single [alternative] user's all connections */
		p_array_init(&ctx.kicks, ctx.who.pool, 1);
		struct kick_session *session = array_append_space(&ctx.kicks);
		session->username = ctx.who.filter.username;
	} else {
		/* Complex kick filter. Iterate all connections and figure out
		   locally which ones to kick. */
		kick_users_get_via_who(&ctx);
	}
	kick_users_via_anvil(&ctx);

	pool_unref(&ctx.who.pool);
}

#define DOVEADM_CMD_KICK_FIELDS \
	.cmd = cmd_kick, \
	.usage = "[-a <anvil socket path>] [-f <passdb field>] [-h <dest host>] <user mask>[|]<ip/bits>", \
DOVEADM_CMD_PARAMS_START \
DOVEADM_CMD_PARAM('a',"socket-path",CMD_PARAM_STR,0) \
DOVEADM_CMD_PARAM('f',"passdb-field",CMD_PARAM_STR,0) \
DOVEADM_CMD_PARAM('h',"dest-host",CMD_PARAM_STR,0) \
DOVEADM_CMD_PARAM('\0',"mask",CMD_PARAM_ARRAY,CMD_PARAM_FLAG_POSITIONAL) \
DOVEADM_CMD_PARAMS_END

struct doveadm_cmd_ver2 doveadm_cmd_kick_ver2 = {
	.name = "kick",
	DOVEADM_CMD_KICK_FIELDS
};
struct doveadm_cmd_ver2 doveadm_cmd_proxy_kick_ver2 = {
	.name = "proxy kick",
	DOVEADM_CMD_KICK_FIELDS
};
